logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java

```diff
@@ -0,0 +1,262 @@
+        this.head = 0;
+        this.pageNum = pageNum;
+        this.capacity = capacity;
+        this.offsetMap = new ArrayList<>();
+        this.checkSummer = new CRC32();
+    }
+
+    // @return the concrete class buffer
+    protected abstract ByteBuffer getBuffer();
+
+    @Override
+    public void open(long minSeqNum, int elementCount) throws IOException {
+        getBuffer().position(0);
+        this.version = getBuffer().get();
+        this.head = 1;
+
+        this.minSeqNum = minSeqNum;
+        this.elementCount = elementCount;
+
+        if (this.elementCount > 0) {
+            // verify first seqNum to be same as expected minSeqNum
+            long seqNum = getBuffer().getLong();
+            if (seqNum != this.minSeqNum) { throw new IOException(String.format("first seqNum=%d is different than minSeqNum=%d", seqNum, this.minSeqNum)); }
+
+            // reset back position to first seqNum
+            getBuffer().position(this.head);
+
+            for (int i = 0; i < this.elementCount; i++) {
+                // verify that seqNum must be of strict + 1 increasing order
+                readNextElement(this.minSeqNum + i, !VERIFY_CHECKSUM);
+            }
+        }
+    }
+
+    // recover will overwrite/update/set this object minSeqNum, capacity and elementCount attributes
+    // to reflect what it recovered from the page
+    @Override
+    public void recover() throws IOException {
+        getBuffer().position(0);
+        this.version = getBuffer().get();
+        this.head = 1;
+
+        // force minSeqNum to actual first element seqNum
+        this.minSeqNum = getBuffer().getLong();
+        // reset back position to first seqNum
+        getBuffer().position(this.head);
+
+        // reset elementCount to 0 and increment to actal number of valid elements found
+        this.elementCount = 0;
+
+        for (int i = 0; ; i++) {
+            try {
+                // verify that seqNum must be of strict + 1 increasing order
+                readNextElement(this.minSeqNum + i, VERIFY_CHECKSUM);
+                this.elementCount += 1;
+            } catch (PageIOInvalidElementException e) {
+                // simply stop at first invalid element
+                break;
+            }
+        }
+
+        // if we were not able to read any element just reset minSeqNum to zero
+        if (this.elementCount <= 0) {
+            this.minSeqNum = 0;
+        }
+    }
+
+    // read and vadidate next element at page head
+    // @param verifyChecksum if true the actual element data will be read + checksumed and compared to written checksum
+    private void readNextElement(long expectedSeqNum, boolean verifyChecksum) throws PageIOInvalidElementException {
+        // if there is no room for the seqNum and length bytes stop here
+        if (this.head + SEQNUM_SIZE + LENGTH_SIZE > capacity) { throw new PageIOInvalidElementException("cannot read seqNum and length bytes past buffer capacity"); }
```

Comment:
I personally prefer it as a one-liner - I really like the fact that this validation code uses less visual real estate and makes the general flow more readable. Again, personal preference.